library/ssl_tls13_server.c

```diff
@@ -24,16 +24,940 @@
+     */
+    mbedtls_ssl_add_hs_msg_to_checksum( ssl, MBEDTLS_SSL_HS_SERVER_HELLO,
+                                        buf, p - buf );
+    /*
+     * Search for a matching ciphersuite
+     */
+    ciphersuites = ssl->conf->ciphersuite_list;
+    ciphersuite_info = NULL;
+    for ( j = 0, p = ciph_offset; j < cipher_suites_len; j += 2, p += 2 )
+    {
+        for ( i = 0; ciphersuites[i] != 0; i++ )
+        {
+            if( p[0] != ( ( ciphersuites[i] >> 8 ) & 0xFF ) ||
+                p[1] != ( ( ciphersuites[i] ) & 0xFF ) )
+                continue;
+
+            ciphersuite_info = mbedtls_ssl_ciphersuite_from_id(
+                               ciphersuites[i] );
+
+            if( ciphersuite_info == NULL )
+            {
+                MBEDTLS_SSL_DEBUG_MSG(
+                1,
+                ( "mbedtls_ssl_ciphersuite_from_id: should never happen" ) );
+                return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+            }
+
+            goto have_ciphersuite;
+
+        }
+    }
+
+    return( MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
+
+have_ciphersuite:
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "selected ciphersuite: %s",
+                                ciphersuite_info->name ) );
+
+    ssl->session_negotiate->ciphersuite = ciphersuites[i];
+    ssl->handshake->ciphersuite_info = ciphersuite_info;
+
+    /* List all the extensions we have received */
+    ssl_tls13_debug_print_client_hello_exts( ssl );
+
+    /*
+     * Determine the key exchange algorithm to use.
+     * There are three types of key exchanges supported in TLS 1.3:
+     * - (EC)DH with ECDSA,
+     * - (EC)DH with PSK,
+     * - plain PSK.
+     *
+     * The PSK-based key exchanges may additionally be used with 0-RTT.
+     *
+     * Our built-in order of preference is
+     *  1 ) Plain PSK Mode
+     *  2 ) (EC)DHE-PSK Mode
+     *  3 ) Certificate Mode
+     */
+
+    if( !ssl_tls13_check_certificate_key_exchange( ssl ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG(
+                1,
+                ( "ClientHello message misses mandatory extensions." ) );
+        MBEDTLS_SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_MISSING_EXTENSION ,
+                                      MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER );
+        return( MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER );
+    }
+
+#if defined(MBEDTLS_SSL_COOKIE_C)
+    /* If we failed to see a cookie extension, and we required it through the
+     * configuration settings ( rr_config ), then we need to send a HRR msg.
+     * Conceptually, this is similiar to having received a cookie that failed
+     * the verification check.
+     */
+    if( ( ssl->conf->rr_config == MBEDTLS_SSL_FORCE_RR_CHECK_ON ) &&
+        !( ssl->handshake->extensions_present & MBEDTLS_SSL_EXT_COOKIE ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG(
+                2,
+                ( "Cookie extension missing. Need to send a HRR." ) );
+        hrr_required = 1;
+    }
+#endif /* MBEDTLS_SSL_COOKIE_C */
```

Comment:
```suggestion
```
Should we remove cookie support at this point? @ronald-cron-arm . 
I'd like reduce the scope of this PR

Reply:
Yes as discussed offline, no cookie support for the time being.